workflows/src/main/kotlin/com/r3/corda/lib/tokens/workflows/internal/selection/VaultWatcherService.kt

```diff
@@ -124,35 +129,43 @@ class VaultWatcherService(tokenObserver: TokenObserver? = null) : SingletonSeria
         }
     }
 
+    // TODO Improve type safety for this method
     @Suspendable
     fun selectTokens(
             owner: Any,
-            amountRequested: Amount<IssuedTokenType>,
+            requiredAmount: Amount<TokenType>,
             predicate: ((StateAndRef<FungibleToken>) -> Boolean) = { true },
             allowShortfall: Boolean = false,
             autoUnlockDelay: Duration = Duration.ofMinutes(5),
             selectionId: String
     ): List<StateAndRef<FungibleToken>> {
-        val set = getTokenBucket(owner, amountRequested.token.tokenType.tokenClass, amountRequested.token.tokenType.tokenIdentifier)
+        // TODO terrible hack for making it work for now, refactor
+        val buckets = if (owner is List<*>) {
+            owner.map { getTokenBucket(it as PublicKey, requiredAmount.token.tokenClass, requiredAmount.token.tokenIdentifier) }
+        } else {
+            listOf(getTokenBucket(owner, requiredAmount.token.tokenClass, requiredAmount.token.tokenIdentifier))
+        }
         val lockedTokens = mutableListOf<StateAndRef<FungibleToken>>()
-        var amountLocked: Amount<IssuedTokenType> = amountRequested.copy(quantity = 0)
-        for (tokenStateAndRef in set.keys) {
-            //does the token satisfy the (optional) predicate?
-            if (amountRequested.token.issuer == tokenStateAndRef.state.data.amount.token.issuer && predicate.invoke(tokenStateAndRef)) {
-                //if so, race to lock the token, expected oldValue = PLACE_HOLDER
-                if (set.replace(tokenStateAndRef, PLACE_HOLDER, selectionId)) {
-                    //we won the race to lock this token
-                    lockedTokens.add(tokenStateAndRef)
-                    val token = tokenStateAndRef.state.data
-                    amountLocked += uncheckedCast(token.amount)
-                    if (amountLocked >= amountRequested) {
-                        break
+        var amountLocked: Amount<TokenType> = requiredAmount.copy(quantity = 0)
+        for (set in buckets) {
```

Comment:
@roastario to review this section.

Reply:
After I raise a pr for external id, as we discussed. Now it doesn't make much sense